File: scala/src/main/scala/free/core/interpreters/Production.scala
package free.core.interpreters

import cats.effect.IO
import free.core.{App, ConsoleDSL, AppDSL, CoinFlipDSL}
import free.core.App.*
import free.core.AppDSL.*
import free.core.CoinFlipDSL.*
import free.core.ConsoleDSL.*
import cats.~>
import free.core.CoinFlipDSL

object Production:
  val productionInterpreter = new (AppDSL ~> IO):
    override def apply[A](a: AppDSL[A]): IO[A] = a match
      case EvalCoinFlip(c) => c.foldMap(productionCoinFlipInterpreter)
      case EvalConsole(c)  => c.foldMap(productionConsoleInterpreter)

  val productionCoinFlipInterpreter = new (CoinFlipDSL ~> IO):
    override def apply[A](cf: CoinFlipDSL[A]): IO[A] = cf match
      case FlipCoin => IO(util.Random.nextBoolean)

  val productionConsoleInterpreter = new (ConsoleDSL ~> IO):
    override def apply[A](c: ConsoleDSL[A]): IO[A] = c match
      case PrintLine(msg) => IO.println(msg)

File: haskell/lib/Free/Core/Interpreters/Production.hs
module Free.Core.Interpreters.Production (appToIOInterpreter) where

import Free.Core (AppDSL (..), CoinFlipDSL (..), ConsoleDSL (..), runWith, type (~>))
import System.Random (randomIO)

-- To interpret a program one has to define how to interpret
-- each step of the possible actions.
-- Since the app is a simple product of Coin and Console its
-- implementation simply uses the single interpreters

appToIOInterpreter :: AppDSL ~> IO
appToIOInterpreter = \case
  EvalCoinFlip c k -> k <$> c `runWith` coinToIOInterpreter
  EvalConsole c k -> k <$> c `runWith` consoleToIOInterpreter

coinToIOInterpreter :: CoinFlipDSL ~> IO
coinToIOInterpreter = \case
  FlipCoin k -> k <$> randomIO

consoleToIOInterpreter :: ConsoleDSL ~> IO
consoleToIOInterpreter = \case
  PrintLine msg k -> k <$> putStrLn msg
